alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+            i, l = j[row.item()], col.item()  # noqa: E741
+
+            # update the score
+            covered = np.sum(delta_xi_all[:, l, B[i].astype(bool)], axis=-1)
+            delta_xi_all[:, l, B[i].astype(bool)] = 0
+            delta_xi_summed[:, l] -= covered
+            scores_all[:, l] -= covered
+
+            # add prototype to the corresponding list according to the class label `l` that it covers
+            # and remove the index `i` from list of available indices
+            protos[l].append(i)
+            available_indices.remove(i)
+
+        return self._build_explanation(protos)
+
+    def _build_explanation(self, protos: Dict[int, List[int]]) -> Explanation:
+        """
+        Helper method to build `Explanation` object.
+        """
+        data = deepcopy(DEFAULT_DATA_PROTOSELECT)
+        data['prototypes_indices'] = np.concatenate(list(protos.values())).astype(np.int32)
+        data['prototypes_labels'] = np.concatenate([[l] * len(protos[l])
+                                                    for l in protos]).astype(np.int32)  # noqa: E741
+        data['prototypes'] = self.Y[data['prototypes_indices']]
+        return Explanation(meta=self.meta, data=data)
+
+    def save(self, path: Union[str, os.PathLike]) -> None:
+        super().save(path)
+
+    @classmethod
+    def load(cls, path: Union[str, os.PathLike]) -> "Explainer":  # type: ignore[override]
+        return super().load(path, predictor=None)
+
+
+def _helper_protoselect_euclidean_1knn(explainer: ProtoSelect,
+                                       num_prototypes: int,
+                                       eps: float,
+                                       preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None
+                                       ) -> Optional[KNeighborsClassifier]:
+    """
+    Helper function to fit a 1-KNN classifier on the prototypes returned by the explainer.
+    Sets the epsilon radius to be used.
+
+    Parameters
+    ----------
+    explainer
+        Fitted explainer.
+    num_prototypes
+        Number of requested prototypes.
+    eps
+        Epsilon radius to be set and used for the computation of prototypes.
+     preprocess_fn
+        Preprocessing function to be applied to the data instance before fitting the 1-KNN classifier
+
+    Returns
+    -------
+    Fitted KNN-classifier.
+    """
+    # update explainer eps and get explanation
+    explainer.eps = eps
+    explanation = explainer.explain(num_prototypes=num_prototypes)
+
+    # train 1-knn classifier
+    proto, proto_labels = explanation.data['prototypes'], explanation.data['prototypes_labels']
+    if len(proto) == 0:
+        return None
+
+    knn = KNeighborsClassifier(n_neighbors=1)
+    return knn.fit(X=proto, y=proto_labels)
+
+
+def cv_protoselect_euclidean(refset: Tuple[np.ndarray, np.ndarray],
+                             protoset: Tuple[np.ndarray, ],
```

Comment:
Could this be `Optional` and if `None` then `X_ref` is used, same as in the `ProtoSelect` implementation?

Reply:
Yes. It can be optional.